### PR TITLE
Rate limit writing of options when queuing sync

### PR DESCRIPTION
**Pull request by gravityrail**
Fixes an issue where writing options too quickly can introduce issues like replication lag on some sites.

### CODE REVIEW

File: tests/php/sync/test_class.jetpack-sync-full.php [hunk 1]
```diff
@@ -825,6 +825,18 @@ function test_full_sync_doesnt_send_deleted_users() {
 		$this->assertEquals( $keep_user_id, $users[ $existing_user_count ]->ID );
 	}
 
+	function test_full_sync_doesnt_exceed_options_write_rate_limit() {
+		foreach( range( 1, 2100 ) as $i ) { // 200 items+
```
Comment (enejb) [inline tests/php/sync/test_class.jetpack-sync-full.php]: should we manually set the setting here? Just so the test is not so brittle?

Comment (gravityrail) [inline tests/php/sync/test_class.jetpack-sync-full.php]: good call - let's do that in a separate PR otherwise we'll have to wait for the tests to run again :/

File: sync/class.jetpack-sync-queue.php [hunk 2]
```diff
@@ -54,6 +54,8 @@ function add( $item ) {
 			) );
 			$added      = ( 0 !== $rows_added );
 		}
+
+		do_action( 'jpsq_item_added' );
```
Comment (ebinnion) [inline sync/class.jetpack-sync-queue.php]: We need documentation for this action. And perhaps it should be name `jetpack_sync_queue_item_added`.

File: tests/php/sync/test_class.jetpack-sync-full.php (hunk 1 above)
Comment (lezama) [inline tests/php/sync/test_class.jetpack-sync-full.php]: 2100 or 210?

Comment (gravityrail) [inline tests/php/sync/test_class.jetpack-sync-full.php]: 2100 = 2 * 100-items-per-second * 10-users-per-queue-item + some margin

**Comment by gravityrail** [comment]
Merged to branch-4.2
